crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -214,76 +213,96 @@ where
             .get_segment_header(last_segment_index)
             .expect("Last segment index exists; qed");
 
-        let second_last_segment_index = last_segment_header.segment_index() - SegmentIndex::ONE;
-
-        let second_last_segment_blocks = download_and_reconstruct_blocks(
-            second_last_segment_index,
+        // Just to seed reconstructor, we don't care about blocks themselves
+        download_and_reconstruct_blocks(
+            last_segment_header.segment_index() - SegmentIndex::ONE,
             self.piece_getter,
             &mut reconstructor,
         )
         .await?;
 
-        let blocks_in_second_last_segment = second_last_segment_blocks.len();
-        debug!(
-            "Second last segment blocks downloaded (SegmentId={}): {}-{}",
-            second_last_segment_index,
-            second_last_segment_blocks[0].0,
-            second_last_segment_blocks[blocks_in_second_last_segment - 1].0
-        );
-
         let blocks = download_and_reconstruct_blocks(
             last_segment_header.segment_index(),
             self.piece_getter,
             &mut reconstructor,
         )
         .await?;
+        let mut blocks = VecDeque::from(blocks);
 
-        let blocks_in_last_segment = blocks.len();
-        debug!(
-            "Blocks downloaded (SegmentId={}): {}-{}",
-            last_segment_index,
-            blocks[0].0,
-            blocks[blocks_in_last_segment - 1].0
-        );
-
-        // 2. Raw import the last block from the second last segment.
+        let (first_block_number, first_block_bytes) = blocks
+            .pop_front()
+            // TODO: Ensure this expectation actually holds, currently this is not guaranteed
+            .expect("List of blocks is not empty according to logic above; qed");
 
-        // 3. Download state for the first block of the last segment.
+        let last_block_number = blocks
```

Comment:
Please, introduce a strict check for the last block number.

Reply:
Not sure what you mean because it is possible that there is just one block to begin with. Once we pop it above `blocks` becomes empty.